numba/core/types/function.py

```diff
@@ -0,0 +1,130 @@
+            self.ftype = ftype
+            name = ftype.name
+        else:
+            raise TypeError(
+                f'constructing {type(self).__name__} from a {type(ftype)}'
+                ' instance is not supported')
+        super(FunctionType, self).__init__(name)
+
+    def signature(self):
+        from numba.core import typing
+        ptype = self.ftype
+        return typing.signature(ptype.rtype, *ptype.atypes)
+
+    @property
+    def key(self):
+        return self.name
+
+    def cast_python_value(self, value):
+        from numba.core import typing
+        if isinstance(value, typing.Signature):
+            ptype = FunctionPrototype(value.return_type, value.args)
+            return FunctionType(ptype)
+        raise NotImplementedError(
+            'cast_python_value({}, {})'.format(value, type(value)))
+
+    def get_call_type(self, context, args, kws):
+        from numba.core import typing
+        from . import Array, Number
+        ptype = self.ftype
+        if len(args) == len(ptype.atypes):
+            for i, atype in enumerate(args):
+                if isinstance(atype, Literal):
+                    atype = atype.literal_type
+                if isinstance(atype, type(ptype.atypes[i])):
+                    continue
+                elif isinstance(atype, Array):
+                    if (
+                            isinstance(ptype.atypes[i], Array)
+                            and atype.dtype <= ptype.atypes[i].dtype):
+                        continue
+                elif isinstance(atype, Number):
+                    if atype <= ptype.atypes[i]:
+                        continue
+                break
+            else:
+                return typing.signature(ptype.rtype, *ptype.atypes)
+            # TODO: implement overload support
+        raise ValueError(
+            f'{self} argument types do not match with {args}')
+
+
+class FunctionPrototype(Type):
+    """
+    Represents a first-class function type.
+    """
+    mutable = True
+    cconv = None
+
+    def __init__(self, rtype, atypes):
+        self.rtype = rtype
+        self.atypes = tuple(atypes)
+
+        assert isinstance(rtype, Type), (rtype)
+        lst = []
+        for atype in self.atypes:
+            assert isinstance(atype, Type), (atype)
+            lst.append(atype.name)
+        name = '%s(%s)' % (rtype, ', '.join(lst))
+
+        super(FunctionPrototype, self).__init__(name)
+
+    @property
+    def key(self):
+        return self.name
+
+
+class WrapperAddressProtocol(ABC):
+    """Base class for Wrapper Address Protocol.
+
+    Objects that inherit from the WrapperAddressProtocol can be passed
+    as arguments to Numba jit compiled functions where it can be used
+    as first-class functions. As a minimum, the derived types must
+    implement the two methods __wrapper_address__ and signature.
+
+    """
+
+    @abstractmethod
+    def __wrapper_address__(self, sig):
```

Comment:
Why does this method need to take a `Signature` argument? In what case would `sig != self.signature`?

Reply:
The WAP object may hold addresses to different functions. See, for example
https://github.com/Quansight/numba/blob/pearu/3405/numba/tests/test_function_type.py#L532-L559

That said, I can see that `sig` argument could be eliminated for `CFunc` objects. Also in the case of WAP objects, the `sig` argument is used only for ensuring signature compatibility that holds always except in possible bugs in user-code. In fact, the example referenced above could be redesigned such that `__wrapper_address__` would not need the `sig` argument.

However, the situation for eliminating `sig` argument is not clear in the case of `Dispatcher` objects (`njit` decorated functions) where the `__wrapper_address__` may need to compile the `sig`-specific function in runtime.

I think this question requires research with respect to supporting `njit` decorated functions that I am currently working on. To enable both ways (until the questions is resolved), we could use
```
@abstractmethod
def __wrapper_address__(self, sig=None):
```
for defining the interface. What do you think?

Reply:
I can now confirm that `sig` is not required for njit-decorated first-class functions.

Reply:
The signature argument is now removed from `__wrapper_address__`.